File: luogu/P_1165_日志分析.cpp
#include <iostream>
#include <vector>

struct Item {
    int id;
    int mass;

    Item(int id, int mass) : id(id), mass(mass) {}
};

int main() {
    std::vector<Item> que = {Item(0, 0)};
    int cnt = 0;

    int n;
    std::cin >> n;

    for (int i = 0; i < n; ++i) {
        int cmd;
        std::cin >> cmd;

        switch (cmd) {
            case 0:
                int mass;
                std::cin >> mass;
                cnt++;
                if (mass > que.back().mass) {
                    que.push_back(Item(cnt, mass));
                }
                break;
            case 1:
                if (que.back().id == cnt) {
                    que.pop_back();
                }
                cnt--;
                break;
            case 2:
                std::cout << que.back().mass << std::endl;
        } 
    }
    return 0;
}

File: luogu/P_1165_日志分析.py
from dataclasses import dataclass


@dataclass
class Item:
    id: int
    mass: int


que: list[Item] = [Item(0, 0)]
cnt = 0
for _ in range(int(input())):
    cmd = tuple(map(int, input().split()))

    if cmd[0] == 0:
        cnt += 1
        mass = cmd[1]
        if mass > que[-1].mass:
            que.append(Item(cnt, mass))
    elif cmd[0] == 1:
        if que[-1].id == cnt:
            que.pop(-1)
        cnt -= 1
    elif cmd[0] == 2:
        print(que[-1].mass)
